only include callcenter datasources that are used by the domain

snopoke opened the pull request:
Currently all 3 UCR datasources are enabled for every Callcenter domain. This updates the config after app builds so that we can only run datasources that are in use in app builds.

snopoke committed:
store enabled datasources in domain callcenter config

snopoke committed:
only include enabled datasources

snopoke committed:
update tests

snopoke committed:
update tests and fix

snopoke committed:
run on app save but only for builds

snopoke committed:
[ci skip] comment

emord reviewed (approved):
I don't have a lot of context on call center, but this seems reasonable. 

It does feel like it adds a bit of complexity here, specifically thinking about the added post-save signal and if there's a bug later on, whether or not it has the data source enabled. but it's probably worth it if we want to keep the queues down

snopoke committed:
catch exceptions for safety

Since this isn't a critical piece of functionality we don't want
to break app saves if this fails.

snopoke committed:
add missing args

millerdev commented on corehq/apps/callcenter/utils.py:
```diff
@@ -23,7 +24,15 @@
+    use_fixtures = attr.ib()
+    form_datasource_enabled = attr.ib()
+    case_datasource_enabled = attr.ib()
+    case_actions_datasource_enabled = attr.ib()
```
suggestion: set defaults so only non-default values are required for construction.

millerdev commented on corehq/apps/domain/models.py:
```diff
@@ -129,12 +129,34 @@ class CallCenterProperties(DocumentSchema):
 
+    form_datasource_enabled = BooleanProperty()
+    case_datasource_enabled = BooleanProperty()
+    case_actions_datasource_enabled = BooleanProperty()
```
Should these have default values?

millerdev commented on corehq/apps/callcenter/tests/test_datasources.py:
```diff
@@ -1,34 +1,27 @@
-            table_ids
-        )
+@generate_cases([
+    (DomainLite('d', None, None, True, True, False, False), ['cc_forms']),
```
Would be nice to use keyword arguments for clarity.
```py
DomainLite('d',
    case_datasource_enabled=False,
    case_actions_datasource_enabled=False)
```

millerdev commented on corehq/apps/callcenter/utils.py:
```diff
@@ -23,7 +24,15 @@
 
 
-class DomainLite(namedtuple('DomainLite', 'name default_timezone cc_case_type use_fixtures')):
+@attr.s
```
❤️

millerdev reviewed (approved):
Few minor comments. Nothing blocking.

snopoke committed:
use kwargs

snopoke committed:
set defaults

snopoke commented:
@millerdev updated

millerdev commented on corehq/apps/userreports/tests/utils.py:
```diff
@@ -101,7 +101,7 @@ def doc_to_change(doc):
 def domain_lite(name):
     from corehq.apps.callcenter.utils import DomainLite
-    return DomainLite(name, None, None, True)
+    return DomainLite(name, None, None, True, True, True, True)
```
nit: this change is unnecessary

millerdev commented on corehq/apps/callcenter/tests/test_utils.py:
```diff
@@ -338,7 +338,7 @@ class DomainTimezoneTests(SimpleTestCase):
     def _test_midnights(self, utcnow, test_cases):
         for tz, offset, expected in test_cases:
-            dom = DomainLite('', tz, '', True)
+            dom = DomainLite('', tz, '', True, True, True, True)
```
nit: this change is unnecessary

millerdev commented on corehq/apps/callcenter/utils.py:
```diff
@@ -23,7 +24,15 @@
+    name = attr.ib()
+    default_timezone = attr.ib()
+    cc_case_type = attr.ib()
+    use_fixtures = attr.ib()
```
nit: Would it also make sense to set defaults (`None, None, True`) for the above three fields? That would allow the `test_call_center_data_source_provider` generated test cases to omit those arguments.

snopoke committed:
remove default args
